### PR TITLE
Messaging Message List

### PR DESCRIPTION
**Pull request by problemv**
For #836 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

https://github.com/openstack/zaqar/blob/8e0dd9763aba67bb7094cf92312daeba8d087866/zaqar/transport/wsgi/v2_0/messages.py#L255

### CODE REVIEW

**Comment by jtopjian** [comment]
@problemv That OpenLab error is ignorable.

You haven't requested a review on this, but I did a quick test. At least in my testing environment, I had to inject a sleep of 30 seconds in order for the messages to retrieve. I'm not sure if that's an issue with my environment configuration or not.

**Comment by problemv** [comment]
@jtopjian I ran the scenario tests without issue in my testing env. I will confirm in the morning again.

**Comment by jtopjian** [comment]
@problemv sounds good -- thanks for checking! It's quite possible it's just my environment. In which case, no changes are needed at all.

**Comment by problemv** [comment]
@jtopjian I didn't want to wait until the morning.

I am able to confirm the tests run within two seconds. 

`stack@problem-VirtualBox:/home/problem$ time gophercloudtest Test messaging/v2
~/testing/go/src/github.com/gophercloud/gophercloud /home/problem
=== RUN   TestListMessages
--- PASS: TestListMessages (0.16s)
	messaging.go:15: Attempting to create Queue: ACPTTESTpZtiG
	messaging.go:49: Attempting to get Queue: ACPTTESTpZtiG

... snip ...

/home/problem

real	0m1.707s
user	0m1.384s
sys	0m0.378s`

**Comment by jtopjian** [comment]
@problemv OK and just to confirm: the test prints out the three messages? That's the problem I ran into: I had to sleep for 30 seconds after message creation in order for messages to be retrieved. Otherwise the queue was empty:

```
2018/04/24 02:02:38 [DEBUG] OpenStack Request URL: GET http://zaqar:8888/v2/queues/ACPTTESTkLHHu/messages
2018/04/24 02:02:38 [DEBUG] OpenStack request Headers:
Accept: application/json
Client-Id: 3381af92-2b9e-11e3-b191-71861300734d
User-Agent: gophercloud/2.0.0
X-Auth-Token: ***
2018/04/24 02:02:39 [DEBUG] OpenStack Response Code: 200
2018/04/24 02:02:39 [DEBUG] OpenStack Response Headers:
Content-Length: 29
Content-Type: application/json; charset=UTF-8
Date: Tue, 24 Apr 2018 02:02:38 GMT
Server: Apache
2018/04/24 02:02:39 [DEBUG] OpenStack Response Body: {
  "links": [],
  "messages": []
}
```

Again, no big deal - I'm more curious about the behavior.

**Comment by problemv** [comment]
@jtopjian Yep here is my full run:

```
stack@problem-VirtualBox:~$ gophercloudtest TestListMessages messaging/v2 --debug
~/testing/go/src/github.com/gophercloud/gophercloud ~
=== RUN   TestListMessages
--- PASS: TestListMessages (0.18s)
	messaging.go:15: Attempting to create Queue: ACPTTEST59KB5
	messaging.go:49: Attempting to get Queue: ACPTTEST59KB5
	messaging.go:34: Created Queue: ACPTTEST59KB5
	messaging.go:71: Attempting to add message to Queue: ACPTTEST59KB5
	messaging.go:83: Successfully added message to queue: ACPTTEST59KB5
	messaging.go:71: Attempting to add message to Queue: ACPTTEST59KB5
	messaging.go:83: Successfully added message to queue: ACPTTEST59KB5
	messaging.go:71: Attempting to add message to Queue: ACPTTEST59KB5
	messaging.go:83: Successfully added message to queue: ACPTTEST59KB5
	tools.go:72: {
		  "body": {
		    "Key": "ACPTTESTMEfM7WC2"
		  },
		  "age": 0,
		  "href": "/v2/queues/ACPTTEST59KB5/messages/5adf644e99352b0509d91bed",
		  "id": "5adf644e99352b0509d91bed",
		  "ttl": 300,
		  "delay": 0,
		  "checksum": ""
		}
	tools.go:72: {
		  "body": {
		    "Key": "ACPTTESTFMz6cyG9"
		  },
		  "age": 0,
		  "href": "/v2/queues/ACPTTEST59KB5/messages/5adf644e99352b0509d91bef",
		  "id": "5adf644e99352b0509d91bef",
		  "ttl": 300,
		  "delay": 0,
		  "checksum": ""
		}
	tools.go:72: {
		  "body": {
		    "Key": "ACPTTESTdZeyC31G"
		  },
		  "age": 0,
		  "href": "/v2/queues/ACPTTEST59KB5/messages/5adf644e99352b050ac6ff6d",
		  "id": "5adf644e99352b050ac6ff6d",
		  "ttl": 300,
		  "delay": 0,
		  "checksum": ""
		}
	messaging.go:39: Attempting to delete Queue: ACPTTEST59KB5
	messaging.go:45: Deleted Queue: ACPTTEST59KB5
PASS
ok  	github.com/gophercloud/gophercloud/acceptance/openstack/messaging/v2	0.181s
~
```

**Comment by problemv** [comment]
@jtopjian in your config do you have a value set for default_message_delay?

**Comment by problemv** [comment]
@jtopjian Yeah that config looks fine to me.

**Comment by jtopjian** [comment]
@problemv Thank you for taking a look :)

Does devstack use redis or mongo? If the latter, I wonder if that's the cause? I used redis simply because I didn't feel like dealing with Mongo.

**Comment by problemv** [comment]
@jtopjian devstack uses Mongo.

**Comment by jtopjian** [comment]
@problemv OK, then we can probably chalk the difference up to something with Redis :)

**Comment by problemv** [comment]
@jtopjian That is weird though I will test Redis tomorrow and see if there is some patch needed for Upstream Zaqar. :)

**Comment by problemv** [comment]
@jtopjian Oh btw this is ready for review :P :)

File: openstack/messaging/v2/messages/results.go [hunk 1]
```diff
@@ -12,6 +15,22 @@ type CreateResult struct {
...
+}
+
+// Message represents a message on a queue.
+type Message struct {
+	Body     map[string]interface{} `json:"body"`
+	Age      int                    `json:"age"`
+	Href     string                 `json:"href"`
+	ID       string                 `json:"id"`
+	TTL      int                    `json:"ttl"`
+	Delay    int                    `json:"delay"`
```
Comment (jtopjian) [inline openstack/messaging/v2/messages/results.go]: I'm not able to find where "delay" is defined as a field. It looks like all of the other fields are correct:

https://github.com/openstack/zaqar/blob/a8215f72f960aeb22f3e3e30abf79b74617def4f/zaqar/api/v2/response.py#L32-L60

Can you verify this?

File: openstack/messaging/v2/messages/results.go [hunk 2]
```diff
@@ -23,3 +42,36 @@ func (r CreateResult) Extract() (ResourceList, error) {
...
+func ExtractMessages(r pagination.Page) ([]Message, error) {
+	var s struct {
+		Messages []Message `json:"messages"`
+	}
+	err := (r.(MessagePage)).ExtractInto(&s)
+	return s.Messages, err
+}
+
+// IsEmpty determines if a MessagePage contains any results.
+func (page MessagePage) IsEmpty() (bool, error) {
```
Comment (jtopjian) [inline openstack/messaging/v2/messages/results.go]: `r MessagePage`

File: openstack/messaging/v2/messages/results.go (hunk 1 above)
Comment (problemv) [reply openstack/messaging/v2/messages/results.go]: You are correct. Sorry about that.

**Comment by problemv** [comment]
@jtopjian Ready to go :)

**Comment by animationzl** [comment]
recheck

**Comment by jtopjian** [comment]
LGTM - thank you :)

**Pull request merged**